File: AddressBook/Java-AddressBook/java_AddressBook_InitialVersion/src/AddressBook/AddressBook.java
package AddressBook;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;
import java.util.Collections;

public class AddressBook
{
    //디폴트생성자
    public AddressBook()
    {
        this.personals = new ArrayList<Personal>();
    }
    //복사생성자
    public AddressBook(AddressBook source)
    {
        //얕은 복사
        //this.personals = (ArrayList<Personal>) source.personals.clone();

        //깊은 복사
        Personal deepCopyPersonal = null;
        //this.personals = new ArrayList<Personal>();
        this.personals = (ArrayList<Personal>)source.personals.clone();

        for(int i = 0; i < this.personals.size(); i++)
        {
            deepCopyPersonal = new Personal(source.getAt(i));
            this.personals.set(i, deepCopyPersonal);
        }
        /*
        for (Personal personal : source.personals)
        {
               deepCopyPersonal = new Personal(personal);
               this.personals.add(deepCopyPersonal);
        }
         */
    }
    //personals 가져오기
    public ArrayList<Personal> getPersonals()
    {
        return this.personals;
    }
    //Personal 객체 정보 기재하기
    public  int record(String name, String address, String telephoneNumber, String emailAddress)
    {
        int index = -1;//기재하기 실패할 경우 -1반환
        Personal personal = new Personal(name, address, telephoneNumber, emailAddress);
        boolean doesRecordingSucceed = this.personals.add(personal);
        if(doesRecordingSucceed == true)
        {
            index = this.personals.indexOf(personal);
        }
        return index;
    }
    //Personal 객체 정보 가져오기
    public Personal getAt(int index)
    {
        return this.personals.get(index);
    }
    //name 으로 Personal 객체 찾기
    public ArrayList<Integer> find(String name)
    {
        //name으로 찾은 Personal 객체들의 위치를 저장할 위치ArrayList 생성하기
        ArrayList<Integer> indexes = new ArrayList<>();
        //주소록의 ArrayList 의 처음부터 끝까지 반복을 돌리면서 name과 같은 배열요소의 위치를 저장한다.
        Personal personal =null;
        for(int i = 0; i < this.personals.size(); i++)
        {
            //해당 위치의 Personal객체를 구한다.
            personal = this.personals.get(i);
            //해당 Personal의 name이 찾고자 하는 name과 같으면
            if(name.compareTo(personal.getName())==0)
            {
                //위치를 저장한다.
                indexes.add(i);
            }
        }
        return indexes;
    }

    /*
    //name 으로 Personal 객체 찾기
    public int[] find(String name)
    {

        //시간 절약 메모리 낭비
        //같은 이름이 없다고 초기화
        boolean isMatched = false;
        //ArrayList 중에서 같은 이름의 배열요소의 위치를 담을 공간
        int[] indexes = new int[this.personals.size()];
        //ArrayList 의 처음부터 끝까지 반복을 돌리면서 찾으려는 이름이 있는지 확인하기
        int index = 0;//배열첨자
        Personal personal =null;
        for(int i = 0; i < this.personals.size(); i++)
        {
            personal = this.personals.get(i);
            if(name.compareTo(personal.getName())==0)
            {
                //위치를 저장한다.
                indexes[index] = i;
                //위치배열 첨자를 증가시킨다.
                index++;
                //찾은게 있다고 표시한다.
                isMatched = true;
            }
        }
        if(isMatched == false)
        {
            indexes = null;
        }
        return indexes;
 */
/*
        //메모리 절약 하지만 시간이 많이 소모됨
        //배열값 초기화
        int[] indexes = null;
        int count = 0;//찾은 개수 초기화
        //ArrayList 의 처음부터 끝까지 찾고자 하는 이름과 비교하여 찾은 개수를 센다.
        for (Personal personal: this.personals)
        {
            if(name.compareTo(personal.getName())==0)
            {
                //찾은 개수를 증가시킨다.
                count++;
            }
        }
        //찾은게 있으면
        if(count > 0)
        {
            //배열을 할당한다.
            indexes = new int[count];
        }
        //ArrayList 의 다시 처음부터 끝까지 반복을 돌리면서 같은 배열요소의 위치를 저장한다.
        int index = 0;//배열첨자
        Personal personal =null;
        for(int i = 0; i < this.personals.size(); i++)
        {
            personal = this.personals.get(i);
            if(name.compareTo(personal.getName())==0)
            {
                //위치를 저장한다.
                indexes[index] = i;
                //배열첨자를 증가시킨다.
                index++;
            }
        }
        return indexes;
    }
    */

    //Personal 객체 정보 수정하기(주소, 전화번호, 이메일주소 변경 가능, 이름은 변경 불가)
    public int correct(int index, String address, String telephoneNumber, String emailAddress)
    {
        //새로운 객체가 생성되기 때문에 이렇게 하면 안됨
        // Personal personal = new Personal(this.personals.get(index).getName(), address,
        //        telephoneNumber, emailAddress);
        Personal personal = this.personals.get(index);
        personal.setAddress(address);
        personal.setTelephoneNumber(telephoneNumber);
        personal.setEmailAddress(emailAddress);
        //this.personals.set(index, personal);
        index = this.personals.indexOf(personal);
        return  index;
    }

    /*
    public Personal correct(int index, String address, String telephoneNumber, String emailAddress)
    {
        //Personal personal = new Personal(this.personals.get(index).getName(),
        //                           address, telephoneNumber, emailAddress);
        Personal personal = this.personals.get(index);
        personal.setAddress(address);
        personal.setTelephoneNumber(telephoneNumber);
        personal.setEmailAddress(emailAddress);
        //Personal oldPersonal = this.personals.set(index, personal);
        //return  oldPersonal;
        return personal;
    }
     */

    //Erase
    public int erase(int index)
    {
        Personal personal = this.personals.remove(index);
        if(personal != null)
        {
            index = -1;
        }
        return index;
    }
    //Arrange
    public void arrange()
    {
        Collections.sort(this.personals, new Comparator<Personal>() {
            @Override
            public int compare(Personal one, Personal other)
            {
                return one.getName().compareTo(other.getName());
            }
        });
    }
    //인스턴스 멤버
    private ArrayList<Personal> personals;
}

File: AddressBook/C-AddressBook/AddressBook20200324/AddressBook20200324/AddressBook.c
#include "AddressBook.h"
#include<stdio.h>
#include<string.h>//strcmp, strcpy
#include<stdlib.h>//calloc, free
#pragma warning(disable:4996)

int main(int argc, char* argv[])
{
	AddressBook addressBook;
	Long index;
	Long(*indexes);
	Long count;

	//메인 테스트 시나리오
	//1.capacity=6을 입력한다.
	Create(&addressBook, 6);
	//2. 홍길동, 서울 중구, 021766710, Hong을 기재한다.
	index = Record(&addressBook, "홍길동", "서울 중구", "021766710", "Hong");
	printf("%s %s %s %s\n", addressBook.personals[index].name, addressBook.personals[index].address,
		addressBook.personals[index].telephoneNumber, addressBook.personals[index].emailAddress);
	//3. 고길동, 서울 성동구, 029575976, Go를 기재한다.
	index = Record(&addressBook, "고길동", "서울 성동구", "029575976", "Go");
	printf("%s %s %s %s\n", addressBook.personals[index].name, addressBook.personals[index].address,
		addressBook.personals[index].telephoneNumber, addressBook.personals[index].emailAddress);
	//4. 홍길동, 인천 연수구, 0313267926, Hong2를 기재한다.
	index = Record(&addressBook, "홍길동", "인천 연수구", "0313267926", "Hong2");
	printf("%s %s %s %s\n", addressBook.personals[index].name, addressBook.personals[index].address,
		addressBook.personals[index].telephoneNumber, addressBook.personals[index].emailAddress);
	//5. 홍길동을 찾는다.
	Find(&addressBook, "홍길동", &indexes, &count);
	index = 0;
	while (index<count)
	{
		printf("%s %s %s %s\n", addressBook.personals[indexes[index]].name,
			addressBook.personals[indexes[index]].address, addressBook.personals[indexes[index]].telephoneNumber,
			addressBook.personals[indexes[index]].emailAddress);
		index++;
	}
	if (indexes!=NULL)
	{
		free(indexes);
	}
	//6. 세 번째 홍길동의 주소와 전화번호, 이메일주소를 고친다.
	index = Correct(&addressBook, 2, "서울 광진구", "027971723", "Hong2");
	printf("%s %s %s %s\n", addressBook.personals[index].name, addressBook.personals[index].address,
		addressBook.personals[index].telephoneNumber, addressBook.personals[index].emailAddress);
	//7. 고길동을 지운다.
	index = Erase(&addressBook, 1);
	if (index == -1)
	{
		printf("지워졌습니다.\n");
	}
	//8. 최길동, 서울 성동구, 023517134, Choi를 기재한다.
	index = Record(&addressBook, "최길동", "서울 성동구", "023517134", "Choi");
	printf("%s %s %s %s\n", addressBook.personals[index].name, addressBook.personals[index].address,
		addressBook.personals[index].telephoneNumber, addressBook.personals[index].emailAddress);
	//9. 나길동, 서울 종로구, 024366751, NA을 기재한다.
	index = Record(&addressBook, "나길동", "서울 종로구", "024366751", "NA");
	printf("%s %s %s %s\n", addressBook.personals[index].name, addressBook.personals[index].address,
		addressBook.personals[index].telephoneNumber, addressBook.personals[index].emailAddress);
	//10. 최길동을 찾는다.
	Find(&addressBook, "최길동", &indexes, &count);
	index = 0;
	while (index < count)
	{
		printf("%s %s %s %s\n", addressBook.personals[indexes[index]].name,
			addressBook.personals[indexes[index]].address, addressBook.personals[indexes[index]].telephoneNumber,
			addressBook.personals[indexes[index]].emailAddress);
		index++;
	}
	if (indexes != NULL)
	{
		free(indexes);
	}
    //11. 네 번째 정길동의 주소와 전화번호, 이메일주소를 고친다.
	index = Correct(&addressBook, 3, "울산 북구", "0529517699", "NA");
	printf("%s %s %s %s\n", addressBook.personals[index].name, addressBook.personals[index].address,
		addressBook.personals[index].telephoneNumber, addressBook.personals[index].emailAddress);
	printf("=============================================================\n");
    //12. 정리한다.
	Arrange(&addressBook);

	index = 0;
	while (index<addressBook.length)
	{
		printf("%s %s %s %s\n", addressBook.personals[index].name, addressBook.personals[index].address,
			addressBook.personals[index].telephoneNumber, addressBook.personals[index].emailAddress);
		index++;
	}
    //13. 할당해제한다.
    Destroy(&addressBook);

	return 0;
}

//Create
void Create(AddressBook* addressBook, Long capacity)
{
	//1. capacity를 입력받는다.
	//2. 개인들 배열을 만든다.
	addressBook->personals = (Personal(*))calloc(capacity, sizeof(Personal));
	//3. 할당량을 정한다.
	addressBook->capacity = capacity;
	//4. 사용량을 정한다.
	addressBook->length = 0;
	//5. 끝내다.
}

//Record
Long Record(AddressBook* addressBook, char(*name), char(*address),
	char(*telephoneNumber), char(*emailAddress))
{
	Long index = -1;
	Personal(*personals)=NULL;
	Long i;
	
	//1. 성명, 주소, 전화번호, 이메일주소를 입력받는다.
	//2. 빈 줄을 찾는다.
	//3. 빈 줄이 없으면
	if (addressBook->length >= addressBook->capacity)
	{
		//3.1 새주소록을 만든다.
		personals = (Personal(*))calloc(addressBook->capacity + 1, sizeof(Personal));
		//3.2 기존 주소록의 내용을 옮겨적는다.
		i = 0;
		while (i<addressBook->length)
		{
			personals[i] = addressBook->personals[i];
			i++;
		}
		//3.3 기존주소록을 지운다.
		if (addressBook->personals != NULL)
		{
			free(addressBook->personals);
		}
		addressBook->personals = personals;
		addressBook->capacity++;
	}
	//4. 빈줄에 적는다.
	index = addressBook->length;
	strcpy(addressBook->personals[index].name, name);
	strcpy(addressBook->personals[index].address, address);
	strcpy(addressBook->personals[index].telephoneNumber, telephoneNumber);
	strcpy(addressBook->personals[index].emailAddress, emailAddress);
	addressBook->length++;
	//5. 위치를 출력한다.
	return index;
	//6. 끝내다.
}

//Find
void Find(AddressBook* addressBook, char(*name), Long* (*indexes), Long* count)
{
	Long i = 0;
	Long j = 0;

	*count = 0;
	//1. 성명을 입력받는다.
	//2. 기재된 내용보다 작거나 같은 동안 반복한다.
	while (i<addressBook->length)
	{
		//2.1 이름이 서로 같으면
		if (strcmp(addressBook->personals[i].name, name)==0)
		{
			//2.1.1 개수를 센다.
			(*count)++;
		}
		i++;
	}
	//3. 개수가 0보다 크면
	if (*count > 0)
	{
		//3.1 위치들 배열을 만든다.
		*indexes = (Long(*))calloc(*count, sizeof(Long));
	}
	//4. 기재된 내용보다 작거나 같은동안 반복한다.
	i = 0;
	while (i<addressBook->length)
	{
		//4.1 이름이 서로 같으면
		if (strcmp(addressBook->personals[i].name, name)==0)
		{
			//4.1.1 위치를 저장한다.
			(*indexes)[j] = i;
			j++;
		}
		i++;
	}
	//5. 위치들과 개수를 출력한다.
	//6. 끝내다.
}

//Correct
Long Correct(AddressBook* addressBook, Long index, char(*address), char(*telephoneNumber), char(*emailAddress))
{
	//1. 위치, 주소, 전화번호, 이메일주소를 입력받는다.
	//2. 위치에 적는다.
	strcpy(addressBook->personals[index].address, address);
	strcpy(addressBook->personals[index].telephoneNumber, telephoneNumber);
	strcpy(addressBook->personals[index].emailAddress, emailAddress);
	//3. 위치를 출력한다.
	return index;
	//4. 끝내다.
}

//Erase
Long Erase(AddressBook* addressBook, Long index)
{
	Personal(*personals)=NULL;
	Long i = 0;
	Long j = 0;
	//1. 위치를 입력받는다.
	//2. 새주소록을 만든다.
	if (addressBook->capacity > 1)
	{
		personals = (Personal(*))calloc(addressBook->capacity - 1, sizeof(Personal));
	}
	//3. 기존주소록의 내용을 옮겨적는다.
	while (i<index)
	{
		personals[j] = addressBook->personals[i];
		j++;
		i++;
	}
	i = index + 1;
	while (i < addressBook->length)
	{
		personals[j] = addressBook->personals[i];
		j++;
		i++;
	}
	//4. 기존주소록을 지운다.
	if (addressBook->personals!=NULL)
	{
		free(addressBook->personals);
		addressBook->personals = NULL;
	}
	if (addressBook->capacity > 1)
	{
		addressBook->personals = personals;
	}
	addressBook->capacity--;
	addressBook->length--;
	index = -1;
	//5. 위치를 출력한다.
	return index;
	//6. 끝내다.
}

//Arrange
void Arrange(AddressBook* addressBook)
{
	Personal personal;
	Long i = 1;
	Long j;
	//1. 오름차순으로 정렬한다.
	while (i<addressBook->length)
	{
		personal = addressBook->personals[i];
		j = i - 1;
		while (j>=0 && strcmp(addressBook->personals[j].name, personal.name)>0)
		{
			addressBook->personals[j + 1] = addressBook->personals[j];
			j--;
		}
		addressBook->personals[j + 1] = personal;
		i++;
	}
	//2. 끝내다.
}

//Destroy

void Destroy(AddressBook* addressBook)
{
	//1. 할당해제한다.
	if (addressBook->personals != NULL)
	{
		free(addressBook->personals);
	}
	//2. 끝내다.
}

